textarea.go

```diff
@@ -0,0 +1,465 @@
+
+// cursorByScreen return position cursor in TextView.buffer coordinate.
+// unit: rune
+func (f TextArea) cursorByScreen() (bufferLine, bufferPosition int) {
+	if len(f.view.index) == 0 {
+		return
+	}
+	x, _, _, _ := f.GetInnerRect()
+	indexLine := f.cursorIndexLine()
+	bufferLine = f.view.index[indexLine].Line
+	bytePos := f.view.index[indexLine].Pos
+
+	// convert from screen grapheme to buffer position in rune
+	buf := f.view.buffer[bufferLine]
+	runePos := len([]rune(buf[:bytePos]))
+	buf = buf[bytePos:]
+
+	// find amount of runes in view graphemes
+	widthOnScreen := f.view.columnOffset + f.cursor.x - x
+	amountRunes := 0
+	for ; ; amountRunes++ {
+		if len([]rune(buf)) <= amountRunes {
+			break
+		}
+		width := stringWidth(string(([]rune(buf))[:amountRunes]))
+		if widthOnScreen == width {
+			break
+		}
+		if widthOnScreen < width {
+			amountRunes--
+			break
+		}
+	}
+	// position in buffer
+	bufferPosition = runePos + amountRunes
+	return
+}
+
+// cursorByBuffer modify position of cursor in according to position in buffers.
+func (f *TextArea) cursorByBuffer(bufferLine, bufferPosition int) {
+	lastIndexLine := f.cursorIndexLine()
+
+	buffers := f.view.buffer
+	if len(buffers) == 0 || len(f.view.index) == 0 {
+		f.cursor.x = 0
+		f.cursor.y = 0
+		return
+	}
+	// correction bufferLine
+	if bufferLine < 0 {
+		bufferLine = 0
+	} else if len(buffers)-1 < bufferLine {
+		bufferLine = len(buffers) - 1
+	}
+	// correction bufferPosition
+	if bufferPosition < 0 {
+		bufferPosition = 0
+	}
+	// find index
+	indexes := f.view.index
+	indexLine := -1 // position in slice indexes
+	indexPos := -1  // amount rune from indexes[i].Pos
+	isOutsideBuffer := false
+	for i := len(indexes) - 1; i >= 0; i-- {
+		if indexes[i].Line != bufferLine {
+			continue
+		}
+		pos := len([]rune(buffers[bufferLine][:indexes[i].Pos]))
+		if pos <= bufferPosition {
+			indexLine = i
+			if size := len([]rune(buffers[bufferLine])); size < bufferPosition {
+				bufferPosition = size
+				isOutsideBuffer = true
+			}
+			indexPos = bufferPosition - pos
+			break
+		}
+	}
+	if indexLine < 0 {
+		// TODO: find that situation
+		indexLine = 0
+	}
+	if indexPos < 0 {
+		// TODO: find that situation
+		indexPos = 0
+	}
+
+	// convert position from indexes to grapheme for cursor
+	var posInGrapheme int
+	{
+		buf := buffers[indexes[indexLine].Line]
+		partBuf := buf[indexes[indexLine].Pos:]
+		part2 := string(([]rune(partBuf))[:indexPos])
+		posInGrapheme = stringWidth(part2)
+	}
+
+	// store last cursor position
+	lastCy := f.cursor.y
+
+	// cursor must be inside screen
+	x, y, width, height := f.GetInnerRect()
+	_ = width
+	f.cursor.x = posInGrapheme + x - f.view.columnOffset
+	f.cursor.y = indexLine + y - f.view.lineOffset
+	if isOutsideBuffer {
+		f.cursor.x++
+	}
+	if y+height <= f.cursor.y {
+		diff := (indexLine - lastIndexLine) - (y + height - lastCy) + 1
+		f.view.lineOffset += diff
+	}
+	if f.cursor.y < y {
+		diff := -((lastIndexLine - indexLine) - lastCy) - 1
```

Comment:
As best I can tell 'lastCy' shouldn't be involved in this calculation at all.  When the textarea fills the screen lastCy will be zero when we hit the top of the screen and need to scroll earlier lines into view.  When the textarea is offset from the top of the screen though, lastCy will be non-zero and this will cause `diff` to get ridiculous values. As a result lineOffset gets set to incorrect values and all your text disappears off the textarea display. Removing 'lastCy' from this calc made scrolling work as expected.